### ISSUE TITLE
🐛 Not following `references` in TS config files

### ISSUE THREAD

**Issue by Jym77** [webpro-nl/knip#779]
### Prerequisites

- [X] I'm using the latest version
- [X] I've read the relevant [documentation](https://knip.dev)
- [X] I've searched for [existing issues](https://github.com/webpro-nl/knip/issues?q=is%3Aissue)
- [X] I've checked the [list of known issues](https://knip.dev/reference/known-issues)
- [X] I've read the [issue reproduction guide](https://knip.dev/guides/issue-reproduction)

### Reproduction url

[Code Sandbox](https://codesandbox.io/p/devbox/mdkscf?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0ur3hgr0006356ir8mt1qfo%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0ur3hgr0002356ihrdhv0w4%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0ur3hgr0004356itu3v6nng%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0ur3hgr0005356id5x963rl%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0ur3hgr0002356ihrdhv0w4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ur3hgr0001356iqt12uy96%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A13%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A13%252C%2522endColumn%2522%253A20%257D%255D%257D%255D%252C%2522id%2522%253A%2522cm0ur3hgr0002356ihrdhv0w4%2522%252C%2522activeTabId%2522%253A%2522cm0ur3hgr0001356iqt12uy96%2522%257D%252C%2522cm0ur3hgr0005356id5x963rl%2522%253A%257B%2522id%2522%253A%2522cm0ur3hgr0005356id5x963rl%2522%252C%2522activeTabId%2522%253A%2522cm0ur3jox000c356i8cvj8j0d%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SETUP_TASKS%2522%252C%2522id%2522%253A%2522cm0ur3jox000c356i8cvj8j0d%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm0ur3hgr0004356itu3v6nng%2522%253A%257B%2522id%2522%253A%2522cm0ur3hgr0004356itu3v6nng%2522%252C%2522activeTabId%2522%253A%2522cm0ur3hgr0003356ix9eaqoel%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0ur3hgr0003356ix9eaqoel%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522NEW_TERMINAL%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522cm0ur3s0e0011356icqrnz0tk%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Reproduction access

- [X] I've made sure the reproduction is publicly accessible

### Description of the issue

I have a TS package with a `src` and `test` directories. Files in `src` are built to `dist` (which is then included in the package's `files` entry). Files in `test` are built there (and not exported). This implies one `tsconfig.json` in each of these directories.

The `src/foo.ts` file creates some internal exports:
```
/** @internal */
export const foo = 1
```
In real life, this is typically used for stuff that I do not want to put in the package's API but I want to be able to test and thus import from the test files.

The test file then include these as `import { foo } from "./dist/foo.js"`. This is the correct import since TS want them to point to ~the resulting file~ the build artefact.

However, knip reports this as unused:
```
Unused exports (1)
foo  unknown  src/foo.ts:2:14
```

When the test file instead `import { bar } from "../src/foo.js"` (which is incorrect and creates a runtime error), knip does not report it anymore.

So, it seems that knip is not correctly linking the source file with the generated code.

---
Note: when building the repro, initially `import { bar } from "../src/foo.js"` was creating a build error. Only after removing the `outDir` option, building, then setting it up again (and cleaning) did I land in this situation (no build error, runtime error).

**Comment by webpro** [comment]
Knip applies some "source mapping" when the config allows so: https://knip.dev/features/monorepos-and-workspaces#source-mapping to analyze only source files.

It's not actually a monorepo feature, it also applies in your case.

> The test file then include these as `import { foo } from "./dist/foo.js"`. This is the correct import since TS want them to point to the resulting file.

This might be the culprit here. In general you don't want to import "resulting" files from tests or source file. Do you mean build artifacts? And maybe you're referring to import specifiers that have the `.js` extension instead of the actual file's `.ts` extension?

**Comment by Jym77** [comment]
> > The test file then include these as `import { foo } from "./dist/foo.js"`. This is the correct import since TS want them to point to the resulting file.
> 
> This might be the culprit here. In general you don't want to import "resulting" files from tests or source file. Do you mean build artifacts? And maybe you're referring to import specifiers that have the `.js` extension instead of the actual file's `.ts` extension?

Yes, I meant the build artefact.

TS requires me to use the `.js` extension in its imports, not `.ts` (and to point to the correct file). I can't claim I really understand why at the deepest level, but there are numerous discussions:

[TS documentation](https://www.typescriptlang.org/docs/handbook/modules/theory.html)
> Recall that when specifying a file extension on a relative module specifier, TypeScript typically [makes you use the output file extension](https://www.typescriptlang.org/docs/handbook/modules/theory.html#typescript-imitates-the-hosts-module-resolution-but-with-types)

https://github.com/microsoft/TypeScript/issues/55346 or https://github.com/microsoft/TypeScript/issues/59597#issuecomment-2284566272

**Comment by webpro** [comment]
> TS requires me to use the `.js` extension in its imports, not `.ts` (and to point to the correct file). I can't claim I really understand why at the deepest level, but there are numerous discussions:

Target the actual existing TS source file, and replace the `.ts` extension with `.js`. The import specifier should not point to the build artefact (which might not even exist yet).

**Comment by Jym77** [comment]
That, at least, does not work when `verbatimModuleSyntax` is true, as [recommended for libraries](https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#im-writing-a-library) (and actually most cases on that page).

In the code box, `test/foo.spec.ts` contains:
```typescript
import { foo } from "../dist/foo.js";
import { bar } from "../src/foo.js";

console.log(foo + bar);
```
Which is left essentially untouched in  `test/foo.spec.js`, thus resulting in a runtime error `Cannot find module '/project/workspace/src/foo.js'` when trying to `node test/foo.spec.js`.

This may also depends on ESM vs CJS (this is ESM), and of `module` (here `Node16`, which is also the recommended setting).

---
From what I understand of [Module specifiers are not transformed](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-specifiers-are-not-transformed), it looks like the import string is never ever transformed, so it should always point to the build artefact even before it exists.
> There is no compiler option that enables transforming, substituting, or rewriting module specifiers. Consequently, module specifiers must be written in a way that works for the code’s target runtime or bundler, and it’s TypeScript’s job to understand those _output_-relative specifiers.

(emphasis not mine)

**Comment by webpro** [comment]
Module specifiers are not transformed **so they still work after compilation**.

**Comment by Jym77** [comment]
Yes, that a bit the point 😅 Not sure where we're getting 😕 

* TS won't touch module specifiers. Ever.
* So, the correct way to import from the `test` directory is `import { foo } from "../dist/foo.js"`, even if that file (or directory) does not exists yet. It will exist post-compilation, and the specifier must absolutely point to the post-compilation location.
* But knip then does not detect that `foo` is used in the tests, and incorrectly report it as an unused export.

Am I missing something? 🤔 


According to [knip source mapping](https://knip.dev/features/monorepos-and-workspaces#source-mapping):
> If src/index.ts exists, Knip will use that file instead of dist/index.js. 

But this is not what I'm seeing since `foo` is reported as unused export. (the example in the doc has a `baseURL` but adding it does not change the behaviour). Even more confusing, the incorrect `import { bar } from "../src/foo.js"` is not reported.

This may be linked to the directories structure I have with a top-level `tsconfig.json`, and extending it in each of `src` and `test` (since I **do not** want `outDir: "dist"` for the tests, in order to keep them out of the distributed package).

**Comment by webpro** [comment]
> Am I missing something? 🤔

Yes, don't import the file at `dist/foo.js` but `src/foo.js`

> But this is not what I'm seeing since `foo` is reported as unused export.

Happy to discuss after we've sorted the first point. But I can imagine there'll be issues since Knip doesn't actively support `references` or `tsconfig.json` files that are not in the workspace root.

**Comment by webpro** [comment]
> But that just plain out doesn't work (from the TS/node point of view).

Maybe it "doesn't work" in the situation you've created.

There's tools like `ts-node`, `tsx`, `ts-jest`, `jiti` etc. to do the compilation during runtime (so you don't need to compile upfront). Or compile the whole thing first using e.g. `tsc` or `swc` and then the result in `dist` is ready to run in Node.js.

Indeed, the file doesn't exist. Perhaps another thing to "show it works": if you use `./foo.js` as import specifier while there's only `foo.ts` you can still opt-click on it and an IDE like VS Code or WebStorm will go to `foo.ts`. This is how TypeScript module resolution works (but Node.js module resolution doesn't support this indeed).

**Comment by webpro** [comment]
Not sure what to tell you. You're saying "I can't claim I really understand why at the deepest level" and then insist on doing things this way.

It's kinda odd that the test files are compiled into the same directory, and then the resulting JS files are used to exercise the compiled source files. Overall this isn't a best practice and a bit of a mess in the CSB repro. Test runners like Vitest, Bun (or even ts-jest) are there to take the burden away in case the test (and source) files are written in TypeScript. There is no need and there should not be a need for a compilation step of test files. So either **everything** should be compiled upfront if you insist on running things directly in Node.js without any dependencies. Or, I'd recommend, use a test runner that supports TypeScript.

**Comment by Jym77** [comment]
> Not sure what to tell you. You're saying "I can't claim I really understand why at the deepest level" and then insist on doing things this way.

The part I don't fully understand is why TS cannot rewrite module specifiers. Which I do not have a choice of doing things any other ways.

> It's kinda odd that the test files are compiled into the same directory, and then the resulting JS files are used to exercise the compiled source files. Overall this isn't a best practice and a bit of a mess in the CSB repro. Test runners like Vitest, Bun (or even ts-jest) are there to take the burden away in case the test (and source) files are written in TypeScript. There is no need and there should not be a need for a compilation step of test files. So either **everything** should be compiled upfront if you insist on running things directly in Node.js without any dependencies. Or, I'd recommend, use a test runner that supports TypeScript.

Well, I've inherited the repo with a simple test setup not using any runners (granted, it was set up in the early days of TS and TS aware test runners might not have been that common back then). Not enough time or resources to re-architect this when it is effectively working (tests are run, my code is exerciced, bugs and regressions are detected), and even gets a few benefits (low startup time for tests, easy to manually run a single test file for super quick iterations).

Anyway, the same situation can be built without tests. I've updated the sandbox to have several directories inside `src` that are then built into separate `dist-*` directories. Arguably a contrived setup, but not a broken one (from TS + node point of view). Maybe not very realistic, I agree, whatever it tries to do might probably be better handled by `exports` field in the manifest file, or through proper workspaces.

In any case, I know have `src/tsconfig.json` that uses `references` to point to the sub-directories. From `src/index.ts`, I need to `import { foo } from "../dist-foo/foo.js"`. This is the only correct module specifier and any other will break at runtime. But knip does incorrectly report `src/foo/foo.ts` as an unused file.

If I replace the import by `import { foo } from "./foo/foo.js"`, which looks correct under the assumption that all of `src` will be built into the same directory without changing the structure, then knip stops complaining. But that module specifier is wrong an will result into a broken JS file that will crash at runtime.

So, knip not following `references` in TS config files can lead to ignoring some of them and thus report false positives.

**Comment by webpro** [comment]
Guess I steered the discussion sideways myself here, yet I'm also not the one to tell how to organize any project. Still, the repro doesn't meet the requirements:

> That’s why it’s required to provide a minimal reproduction. This contains only the source code and configuration required to demonstrate the issue

I'm going to close this one as it's a bit convoluted. I'm happy to discuss the actual issue at hand in a new issue/repro.
